TestModelZooServer/TestUsingModelZooModel fail when using refactorized codegen

Based on https://github.com/sql-machine-learning/sqlflow/pull/2970 .

```
2020-10-10T06:12:35.8540517Z [0m[91m--- FAIL: TestModelZooServer (4.58s)
2020-10-10T06:12:35.8543005Z [0m[32m    --- PASS: TestModelZooServer/ReleaseModelDef (3.56s)
2020-10-10T06:12:35.8545311Z [0m[0m    modelzooserver_test.go:152: 
2020-10-10T06:12:35.8546656Z [0m[0m        	Error Trace:	modelzooserver_test.go:152
2020-10-10T06:12:35.8548602Z [0m[0m        	Error:      	rpc error: %v
2020-10-10T06:12:35.8549344Z [0m[0m        	Test:       	TestModelZooServer
2020-10-10T06:12:35.8550771Z [0m[0m        	Messages:   	rpc error: code = Unknown desc = convert length head error: strconv.ParseInt: parsing "\x1f\x8b\b\x00\xd3P\x81_\x00\x03": invalid syntax
2020-10-10T06:12:35.8553259Z [0m[91m    --- FAIL: TestModelZooServer/ReleaseTrainedModel (0.01s)
2020-10-10T06:12:35.8556005Z [0m[0m        testing.go:864: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
2020-10-10T06:12:35.8558480Z [0m[0m=== RUN   TestUsingModelZooModel
2020-10-10T06:12:35.8613889Z [0m[0mtime="2020-10-10T14:12:35+08:00" level=info msg="SQLFlow Model Zoo started at: 50056"
2020-10-10T06:12:51.3255821Z [0m[91m--- FAIL: TestUsingModelZooModel (15.47s)
2020-10-10T06:12:51.3256634Z [0m[0m    using_model_test.go:189: 
2020-10-10T06:12:51.3257174Z [0m[0m        	Error Trace:	using_model_test.go:189
2020-10-10T06:12:51.3257869Z [0m[0m        	Error:      	Received unexpected error:
2020-10-10T06:12:51.3258829Z [0m[0m        	            	rpc error: code = Unknown desc = runSQLProgram error: open: table localhost:50056/sqlflow_models.modelzoo_model_iris doesn't exist
2020-10-10T06:12:51.3259747Z [0m[0m        	Test:       	TestUsingModelZooModel
2020-10-10T06:12:51.3260324Z [0m[91mFAIL
2020-10-10T06:12:51.3261771Z [0m[0mcoverage: 47.7% of statements
2020-10-10T06:12:51.3286625Z [0m[91mFAIL	sqlflow.org/sqlflow/go/modelzooserver	21.405s
```

--- a/go/model/model.go
+++ b/go/model/model.go
@@ -229,6 +229,45 @@ func (m *Model) saveDB(connStr, table string, session *pb.Session) (e error) {
 	return nil
 }
 
+// SaveDBExperimental save the model to database with metadata using the refactored format.
+func (m *Model) SaveDBExperimental(connStr, table string, session *pb.Session) (e error) {
+	db, err := database.OpenAndConnectDB(connStr)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	sqlf, e := sqlfs.Create(db, table, session)
+	if e != nil {
+		return fmt.Errorf("cannot create sqlfs file %s: %v", table, e)
+	}
+	defer sqlf.Close()
+
+	metaJSONStr, err := m.Meta.Encode()
+	if err != nil {
+		return err
+	}
+	metaLen := len(metaJSONStr)
+	metaLenHex := fmt.Sprintf("0x%08x", metaLen)
+	sqlf.Write([]byte(metaLenHex))
+	sqlf.Write([]byte(metaJSONStr))
+
+	// model and its metadata are both zipped into a tarball
+	cmd := exec.Command("tar", "czf", "-", "-C", m.workDir, ".")
+	cmd.Stdout = sqlf
+	var errBuf bytes.Buffer
+	cmd.Stderr = &errBuf
+
+	if e := cmd.Run(); e != nil {
+		return fmt.Errorf("tar stderr: %v\ntar cmd %v", errBuf.String(), e)
+	}
+
+	if e := sqlf.Close(); e != nil {
+		return fmt.Errorf("close sqlfs error: %v", e)
+	}
+	return nil
+}
+
 func (m *Model) saveTar(modelDir, save string) (string, error) {
 	save = strings.TrimSuffix(save, ".tar.gz")
 	modelFile := filepath.Join(modelDir, save+".tar.gz")

--- a/go/modelzooserver/modelzooserver_test.go
+++ b/go/modelzooserver/modelzooserver_test.go
@@ -136,10 +136,17 @@ func TestModelZooServer(t *testing.T) {
 		sampleModel := model.New(dir, "SAMPLE TRAIN SELECT")
 		sampleModel.Meta = jsonMeta
 		modelTableName := "sqlflow_models.model_zoo_sample_model"
-		err = sampleModel.Save(modelTableName, &pb.Session{
-			DbConnStr: dbConnStr,
-		})
-		a.NoError(err)
+		if os.Getenv("SQLFLOW_USE_EXPERIMENTAL_CODEGEN") == "true" {
+			err = sampleModel.SaveDBExperimental(dbConnStr, modelTableName, &pb.Session{
+				DbConnStr: dbConnStr,
+			})
+			a.NoError(err)
+		} else {
+			err = sampleModel.Save(modelTableName, &pb.Session{
+				DbConnStr: dbConnStr,
+			})
+			a.NoError(err)
+		}
 
 		req := &pb.ReleaseModelRequest{
 			Name:        modelTableName,

--- a/go/modelzooserver/modelzooserver_test.go
+++ b/go/modelzooserver/modelzooserver_test.go
@@ -136,7 +136,9 @@ func TestModelZooServer(t *testing.T) {
 		sampleModel := model.New(dir, "SAMPLE TRAIN SELECT")
 		sampleModel.Meta = jsonMeta
 		modelTableName := "sqlflow_models.model_zoo_sample_model"
-		if os.Getenv("SQLFLOW_USE_EXPERIMENTAL_CODEGEN") == "true" {
+		// NOTE(typhoonzero): change to if os.Getenv("SQLFLOW_USE_EXPERIMENTAL_CODEGEN") == "true"
+		// when switch to refactored code.
+		if false {
 			err = sampleModel.SaveDBExperimental(dbConnStr, modelTableName, &pb.Session{
 				DbConnStr: dbConnStr,
 			})

